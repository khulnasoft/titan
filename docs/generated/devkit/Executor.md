# Type alias: Executor\<T\>

Ƭ **Executor**\<`T`\>: (`options`: `T`, `context`: [`ExecutorContext`](../../devkit/documents/ExecutorContext)) => `Promise`\<\{ `success`: `boolean` }\> \| `AsyncIterableIterator`\<\{ `success`: `boolean` }\>

Implementation of a target of a project

#### Type parameters

| Name | Type  |
| :--- | :---- |
| `T`  | `any` |

#### Type declaration

▸ (`options`, `context`): `Promise`\<\{ `success`: `boolean` }\> \| `AsyncIterableIterator`\<\{ `success`: `boolean` }\>

##### Parameters

| Name      | Type                                                        |
| :-------- | :---------------------------------------------------------- |
| `options` | `T`                                                         |
| `context` | [`ExecutorContext`](../../devkit/documents/ExecutorContext) |

##### Returns

`Promise`\<\{ `success`: `boolean` }\> \| `AsyncIterableIterator`\<\{ `success`: `boolean` }\>
