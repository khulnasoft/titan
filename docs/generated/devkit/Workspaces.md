# Class: Workspaces

**`Deprecated`**

This will be removed in v19. Use [readProjectsConfigurationFromProjectGraph](../../devkit/documents/readProjectsConfigurationFromProjectGraph) instead.

## Table of contents

### Constructors

- [constructor](../../devkit/documents/Workspaces#constructor)

### Methods

- [readWorkspaceConfiguration](../../devkit/documents/Workspaces#readworkspaceconfiguration)

## Constructors

### constructor

• **new Workspaces**(`root`): [`Workspaces`](../../devkit/documents/Workspaces)

#### Parameters

| Name   | Type     |
| :----- | :------- |
| `root` | `string` |

#### Returns

[`Workspaces`](../../devkit/documents/Workspaces)

## Methods

### readWorkspaceConfiguration

▸ **readWorkspaceConfiguration**(): [`ProjectsConfigurations`](../../devkit/documents/ProjectsConfigurations) & [`NxJsonConfiguration`](../../devkit/documents/NxJsonConfiguration)\<`string`[] \| `"*"`\>

#### Returns

[`ProjectsConfigurations`](../../devkit/documents/ProjectsConfigurations) & [`NxJsonConfiguration`](../../devkit/documents/NxJsonConfiguration)\<`string`[] \| `"*"`\>

**`Deprecated`**

Use [readProjectsConfigurationFromProjectGraph](../../devkit/documents/readProjectsConfigurationFromProjectGraph) instead.
