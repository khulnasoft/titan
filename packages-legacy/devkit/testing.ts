export * from '@titan/devkit/testing';
