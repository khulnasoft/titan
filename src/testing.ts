export {cold, hot} from 'jasmine-marbles';
export {readAll} from './utils/testing';
