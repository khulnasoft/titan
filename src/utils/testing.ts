import {Observable} from 'rxjs/Observable';
import {toArray} from 'rxjs/operator/toArray';
import {toPromise} from 'rxjs/operator/toPromise';

export function readAll<T>(o: Observable<T>): Promise<T[]> {
  return toPromise.call(toArray.call(o));
}
