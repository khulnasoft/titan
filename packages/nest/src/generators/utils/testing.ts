import { addProjectConfiguration, Tree } from '@titan/devkit';
import { createTreeWithEmptyWorkspace } from '@titan/devkit/testing';

export function createTreeWithNestApplication(appName: string): Tree {
  const tree = createTreeWithEmptyWorkspace();
  addProjectConfiguration(tree, appName, {
    root: `${appName}`,
    sourceRoot: `${appName}/src`,
    projectType: 'application',
    targets: {},
  });
  return tree;
}
