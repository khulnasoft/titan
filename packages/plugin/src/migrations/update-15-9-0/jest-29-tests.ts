import {
  createProjectGraphAsync,
  formatFiles,
  readProjectConfiguration,
  Tree,
  visitNotIgnoredFiles,
} from '@nx/devkit';
import { JestExecutorOptions } from '@nx/jest/src/executors/jest/schema';
import { TEST_FILE_PATTERN } from '@nx/jest/src/utils/ast-utils';
import { forEachExecutorOptionsInGraph } from '@nx/devkit/src/generators/executor-options-utils';
import { tsquery } from '@phenomnomnominal/tsquery';
import {
  CallExpression,
  ImportDeclaration,
  VariableStatement,
} from 'typescript';

// NOTE: this is a copy of the @nrwl/jest v15.8.0 migrations
export async function updateTestsJest29(tree: Tree) {
  const graph = await createProjectGraphAsync();
  forEachExecutorOptionsInGraph<JestExecutorOptions>(
    graph,
    '@nrwl/nx-plugin:e2e',
    (options, projectName) => {
      const projectConfig = readProjectConfiguration(tree, projectName);
      visitNotIgnoredFiles(
        tree,
        projectConfig.sourceRoot || projectConfig.root,
        (file) => {
          if (!TEST_FILE_PATTERN.test(file)) {
            return;
          }
          updateJestMockTypes(tree, file);
          updateJestMocked(tree, file);
        }
      );
    }
  );
  await formatFiles(tree);
}

export function updateJestMockTypes(tree: Tree, filePath: string) {
  const contents = tree.read(filePath, 'utf-8');
  const updatedContent = tsquery.replace(
    contents,
    ':matches(ImportDeclaration, VariableStatement):has(Identifier[name="MaybeMockedDeep"], Identifier[name="MaybeMocked"]):has(StringLiteral[value="jest-mock"])',
    (node: ImportDeclaration | VariableStatement) => {
      const text = node.getText();
      return (
        text
          // MaybeMockedDeep and MaybeMocked now are exported as Mocked and MockedShallow
          .replace('MaybeMockedDeep', 'Mocked')
          .replace('MaybeMocked', 'MockedShallow')
      );
    }
  );
  tree.write(filePath, updatedContent);
}

export function updateJestMocked(tree: Tree, filePath: string) {
  const contents = tree.read(filePath, 'utf-8');
  const jestGlobalNodes = tsquery.query(
    contents,
    ':matches(ImportDeclaration, VariableStatement):has(Identifier[name="jest"]):has(StringLiteral[value="@jest/globals"])'
  );

  // this only applies if using jest from @jest/globals
  if (jestGlobalNodes.length === 0) {
    return;
  }

  const updatedJestMockTypes = tsquery.replace(
    contents,
    'CallExpression:has(Identifier[name="jest"]):has(Identifier[name="mocked"])',
    (node: CallExpression) => {
      if (
        node.arguments.length === 2 &&
        node.getText().startsWith('jest.mocked(')
      ) {
        const text = node.getText();
        // jest.mocked(someObject, true); => jest.mocked(someObject);
        if (node.arguments[1].getText() === 'true') {
          return text.replace(/,\s*true/g, '');
        }
        // jest.mocked(someObject, false); => jest.mocked(someObject, {shallow: true});
        // opt into the new behavior unless explicitly opting out
        if (node.arguments[1].getText() === 'false') {
          return text.replace('false', '{shallow: true}');
        }
      }
    }
  );

  tree.write(filePath, updatedJestMockTypes);
}

export default updateTestsJest29;
