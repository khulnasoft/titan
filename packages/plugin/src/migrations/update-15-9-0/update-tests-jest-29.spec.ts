import {
  ProjectGraph,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { libraryGenerator } from '@nx/js';
import { updateTestsJest29 } from './jest-29-tests';

let projectGraph: ProjectGraph;
jest.mock('@nx/devkit', () => ({
  ...jest.requireActual<any>('@nx/devkit'),
  createProjectGraphAsync: jest
    .fn()
    .mockImplementation(async () => projectGraph),
}));
describe('Nx Plugin Migration - jest 29 mocked usage in tests', () => {
  let tree: Tree;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it('should not update anything if there are no tests', async () => {
    await setup(tree, 'my-lib');
    const expected = tree.read('libs/my-lib/src/lib/my-lib.spec.ts', 'utf-8');
    await updateTestsJest29(tree);
    expect(
      tree.read('libs/my-lib/src/file-one.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('libs/my-lib/src/file-two.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('libs/my-lib/src/lib/my-lib.spec.ts', 'utf-8')).toEqual(
      expected
    );
  });
  it('should be idempotent', async () => {
    await setup(tree, 'my-lib');

    const expected = tree.read('libs/my-lib/src/lib/my-lib.spec.ts', 'utf-8');

    await updateTestsJest29(tree);

    expect(
      tree.read('libs/my-lib/src/file-one.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('libs/my-lib/src/file-two.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('libs/my-lib/src/lib/my-lib.spec.ts', 'utf-8')).toEqual(
      expected
    );

    await updateTestsJest29(tree);

    expect(
      tree.read('libs/my-lib/src/file-one.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('libs/my-lib/src/file-two.spec.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(tree.read('libs/my-lib/src/lib/my-lib.spec.ts', 'utf-8')).toEqual(
      expected
    );
  });
});

async function setup(tree: Tree, name: string) {
  await libraryGenerator(tree, {
    name,
  });
  const projectConfig = readProjectConfiguration(tree, name);
  projectConfig.targets['test'] = {
    ...projectConfig.targets['test'],
    configurations: {
      ci: {
        ci: true,
      },
      other: {
        jestConfig: `libs/${name}/jest.config.js`,
      },
    },
  };

  updateProjectConfiguration(tree, name, projectConfig);

  tree.write(
    `libs/${name}/jest.config.js`,
    `module.exports = {
transform: {
  '^.+\\\\.[tj]sx?$': 'ts-jest'
},
moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'html'],
globals: {
    'ts-jest': {
        tsconfig: '<rootDir>/tsconfig.spec.json'
    }
  },
displayName: 'jest',
testEnvironment: 'node',
preset: '../../jest.preset.js'
};

`
  );
  tree.write(
    `libs/${name}/src/file-one.spec.ts`,
    `
import{ MaybeMockedDeep, MaybeMocked } from 'jest-mock';
import {expect, jest, test} from '@jest/globals';
import {song} from './song';

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
  `
  );
  tree.write(
    `libs/${name}/src/file-two.spec.ts`,
    `
const { MaybeMockedDeep, MaybeMocked } = require('jest-mock');
const {expect, jest, test} = require('@jest/globals');
const {song} = require('./song');

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
`
  );
  projectGraph = {
    dependencies: {},
    nodes: {
      [name]: {
        name,
        type: 'lib',
        data: projectConfig,
      } as any,
    },
  };
}
