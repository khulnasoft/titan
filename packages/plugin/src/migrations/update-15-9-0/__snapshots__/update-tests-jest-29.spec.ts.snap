// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nx Plugin Migration - jest 29 mocked usage in tests should be idempotent 1`] = `
"import { MaybeMockedDeep, MaybeMocked } from 'jest-mock';
import { expect, jest, test } from '@jest/globals';
import { song } from './song';

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
"
`;

exports[`Nx Plugin Migration - jest 29 mocked usage in tests should be idempotent 2`] = `
"const { MaybeMockedDeep, MaybeMocked } = require('jest-mock');
const { expect, jest, test } = require('@jest/globals');
const { song } = require('./song');

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
"
`;

exports[`Nx Plugin Migration - jest 29 mocked usage in tests should be idempotent 3`] = `
"import { MaybeMockedDeep, MaybeMocked } from 'jest-mock';
import { expect, jest, test } from '@jest/globals';
import { song } from './song';

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
"
`;

exports[`Nx Plugin Migration - jest 29 mocked usage in tests should be idempotent 4`] = `
"const { MaybeMockedDeep, MaybeMocked } = require('jest-mock');
const { expect, jest, test } = require('@jest/globals');
const { song } = require('./song');

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
"
`;

exports[`Nx Plugin Migration - jest 29 mocked usage in tests should not update anything if there are no tests 1`] = `
"import { MaybeMockedDeep, MaybeMocked } from 'jest-mock';
import { expect, jest, test } from '@jest/globals';
import { song } from './song';

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
"
`;

exports[`Nx Plugin Migration - jest 29 mocked usage in tests should not update anything if there are no tests 2`] = `
"const { MaybeMockedDeep, MaybeMocked } = require('jest-mock');
const { expect, jest, test } = require('@jest/globals');
const { song } = require('./song');

jest.mock('./song');
jest.spyOn(console, 'log');

const mockedSong = jest.mocked(song, true);
// or through \`jest.Mocked<Source>\`
// const mockedSong = song as jest.Mocked<typeof song>;

test('deep method is typed correctly', () => {
  mockedSong.one.more.time.mockReturnValue(12);

  expect(mockedSong.one.more.time(10)).toBe(12);
  expect(mockedSong.one.more.time.mock.calls).toHaveLength(1);
});

test('direct usage', () => {
  jest.mocked(console.log).mockImplementation(() => {
    return;
  });

  console.log('one more time');

  expect(jest.mocked(console.log, false).mock.calls).toHaveLength(1);
});
"
`;
